File: frontend/src/lib/utils/clientIp.ts
import type { ConnectionsResponse } from '$lib/types';

/**
 * Collapses an IPv4-mapped IPv6 address ("::ffff:203.0.113.7") to the plain IPv4
 * it stands for, and leaves everything else alone.
 *
 * Inbounds bind a dual-stack socket, so sing-box reports an IPv4 client through
 * the Clash API in the mapped form. Left as-is it is a DIFFERENT string from the
 * same client's plain address: the connections monitor groups one device under
 * two headings and its saved name never matches (#71). Mirrors
 * util.CanonicalClientIP on the backend, which does the same to the client list
 * and the traffic history.
 */
export function canonicalClientIp(ip: string): string {
	const m = /^::ffff:(\d{1,3}(?:\.\d{1,3}){3})$/i.exec(ip);
	if (!m) return ip;
	return m[1].split('.').every((o) => Number(o) <= 255) ? m[1] : ip;
}

/**
 * The single place connection data is normalised on the way in, so no consumer
 * downstream — grouping key, name lookup, search, breakdown — has to remember to.
 */
export function canonicalizeConnections(resp: ConnectionsResponse): ConnectionsResponse {
	if (!resp?.connections) return resp;
	return {
		...resp,
		connections: resp.connections.map((c) =>
			c.metadata?.sourceIP
				? { ...c, metadata: { ...c.metadata, sourceIP: canonicalClientIp(c.metadata.sourceIP) } }
				: c
		)
	};
}

/**
 * Whether an address can be a device of this box: the LAN, a tunnel, or the box
 * itself. Mirrors util.IsLocalClientIP on the backend, which decides the same
 * for the traffic history and the client roster (#102) — the live Breakdown
 * reads the Clash stream directly, so without this the page contradicted its own
 * historical ranges: a Google front-end that vanishes when you click "1h".
 *
 * IPv6 goes the other way on purpose, exactly as the backend does: a client's
 * IPv6 address is globally routable by design, so the address alone cannot tell
 * a device from a stranger and refusing one would hide a real device.
 *
 * One difference from the backend, which has the settings at hand: a tunnel
 * subnet configured outside the private ranges is not known here, so such a peer
 * is missing from the live view until the range switcher is used. The AWG
 * default is 10.10.0.0/24.
 */
export function isLocalClientIp(ip: string): boolean {
	const addr = canonicalClientIp(ip);
	const v4 = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.exec(addr);
	if (!v4) return addr.includes(':') && !/^(ff|::$)/i.test(addr);
	const [a, b] = v4.slice(1).map(Number);
	if (v4.slice(1).some((o) => Number(o) > 255)) return false;
	return (
		a === 10 ||
		a === 127 ||
		(a === 192 && b === 168) ||
		(a === 172 && b >= 16 && b <= 31) ||
		(a === 169 && b === 254) ||
		(a === 100 && b >= 64 && b <= 127) // RFC 6598
	);
}

/**
 * Drops connections whose source is not a device of this box, keeping the ones
 * with no source at all — those are the box's own dials, shown as "unknown".
 */
export function localSourceConnections<T extends { metadata?: { sourceIP?: string } }>(
	conns: T[]
): T[] {
	return conns.filter((c) => !c.metadata?.sourceIP || isLocalClientIp(c.metadata.sourceIP));
}

File: backend/internal/util/clientip.go
package util

import (
	"net/netip"
	"strings"
)

// CanonicalClientIP collapses an IPv4-mapped IPv6 address ("::ffff:203.0.113.7")
// to the plain IPv4 it stands for, and leaves everything else alone.
//
// Inbounds bind a dual-stack socket, so sing-box reports an IPv4 client through
// the Clash API in the mapped form. Left as-is it is a DIFFERENT string from the
// same client's plain address: the client list grows a second entry that has to
// be named again, traffic history splits into two buckets for one device, and the
// connections monitor shows "::ffff:77.94…" where a name was expected (#71).
//
// Anything unparseable is returned untouched — this normalises a display and
// lookup key, it is not a validator.
func CanonicalClientIP(ip string) string {
	addr, err := netip.ParseAddr(ip)
	if err != nil {
		return ip
	}
	if addr.Is4In6() {
		return addr.Unmap().String()
	}
	return ip
}

// cgnat is RFC 6598 100.64.0.0/10 — the shared address space some tunnel
// setups hand out to peers. netip has no helper for it, hence the constant.
var cgnat = netip.MustParsePrefix("100.64.0.0/10")

// ParsePrefixes turns configured CIDR strings into prefixes, skipping anything
// unparseable. Feeding IsLocalClientIP a setting is then one call, and a typo in
// the settings file costs that one prefix rather than an error path in every
// caller.
func ParsePrefixes(list ...string) []netip.Prefix {
	var out []netip.Prefix
	for _, s := range list {
		if p, err := netip.ParsePrefix(strings.TrimSpace(s)); err == nil {
			out = append(out, p.Masked())
		}
	}
	return out
}

// IsLocalClientIP reports whether ip can be a client of this box: an address
// from the LAN, a tunnel, or the box itself. Public IPv4 addresses are not — in
// router mode they show up in the Clash connection list anyway (#102), where
// they became a "client" row for a Google front-end nobody can name.
//
// extra widens the answer with prefixes this box knows are its own. The AWG
// tunnel subnet is configurable and only defaults to 10.10.0.0/24: an operator
// who picked something globally routable still has peers, and without this their
// traffic would vanish from the peer pages, which read it by tunnel IP.
//
// IPv6 goes the other way on purpose. A LAN client's IPv6 address is globally
// routable by design — a delegated prefix is normal, ULA is the exception — so
// the address alone cannot tell a device from a stranger, and guessing wrong
// erases a real device's history. Only the addresses that can't be a client at
// all (multicast, unspecified) are refused.
//
// Anything unparseable is not local: the value came from a connection's
// sourceIP, and a non-address there is not a device either.
func IsLocalClientIP(ip string, extra ...netip.Prefix) bool {
	addr, err := netip.ParseAddr(CanonicalClientIP(ip))
	if err != nil {
		return false
	}
	for _, p := range extra {
		if p.Contains(addr) {
			return true
		}
	}
	if !addr.Is4() {
		return !addr.IsMulticast() && !addr.IsUnspecified()
	}
	return addr.IsPrivate() || addr.IsLoopback() || addr.IsLinkLocalUnicast() || cgnat.Contains(addr)
}
